tests/integration/test_parallel_replicas_insert_select/test.py

```diff
@@ -0,0 +1,188 @@
+        pytest.param("test_1_shard_3_replicas", 2, True, 2),
+        pytest.param("test_1_shard_3_replicas", 3, False, 4),
+        pytest.param("test_1_shard_3_replicas", 3, True, 3),
+        pytest.param("test_1_shard_3_replicas_1_unavailable", 3, False, 3),
+        pytest.param("test_1_shard_3_replicas_1_unavailable", 3, True, 2),
+        pytest.param("test_1_shard_3_replicas_1_unavailable", 2, False, 2),
+        pytest.param("test_1_shard_3_replicas_1_unavailable", 2, True, 2),
+    ],
+)
+def test_insert_select(start_cluster, cluster_name, max_parallel_replicas, local_pipeline, executed_queries):
+    populate_count = 1000000
+
+    source_table = "t_source"
+    create_tables(source_table, populate_count=populate_count, skip_last_replica=False)
+    target_table = "t_target"
+    create_tables(target_table, populate_count=0, skip_last_replica=False)
+
+    log_comment = uuid.uuid4()
+    node1.query(
+        f"INSERT INTO {target_table} SELECT * FROM {source_table}",
+        settings={
+            "parallel_distributed_insert_select": 2,
+            "enable_parallel_replicas": 2,
+            "max_parallel_replicas": max_parallel_replicas,
+            "cluster_for_parallel_replicas": cluster_name,
+            "log_comment": log_comment,
+            "parallel_replicas_insert_select_local_pipeline": local_pipeline,
+            "enable_analyzer": 1,
+        },
+    )
+    assert (
+        node1.query(
+            f"select count() from {target_table}"
+        )
+        == f"{populate_count}\n"
+    )
+    assert (
+        node1.query(
+            f"select * from {target_table} order by key except select * from {source_table} order by key",
+        )
+        == ""
+    )
+
+    execute_on_cluster(f"SYSTEM FLUSH LOGS query_log")
+    number_of_queries = node1.query(
+            f"""SELECT count() FROM clusterAllReplicas({cluster_name}, system.query_log) WHERE current_database = currentDatabase() AND log_comment = '{log_comment}' AND type = 'QueryFinish' AND query_kind = 'Insert'
+                settings skip_unavailable_shards=1"""
+        )
+
+    if max_parallel_replicas < 3:
+        if local_pipeline:
+            assert(number_of_queries == f"{executed_queries}\n" or number_of_queries == f"{executed_queries-1}\n")
+        else:
+            assert(number_of_queries == f"{executed_queries+1}\n" or number_of_queries == f"{executed_queries}\n")
+    else:
+        assert(number_of_queries == f"{executed_queries}\n")
+
+
+# TODO: Change protocol so we can check if all tables are present on a node
+#       If not, then we can skip such nodes
+#       Currently, we'll just fail
+
+@pytest.mark.parametrize(
+    "cluster_name,max_parallel_replicas,local_pipeline",
+    [
+        pytest.param("test_1_shard_3_replicas", 3, False),
+        pytest.param("test_1_shard_3_replicas", 3, True),
+    ],
+)
+def test_insert_select_no_table(start_cluster, cluster_name, max_parallel_replicas, local_pipeline):
```

Comment:
Looks like this can be a regular stateless test?

Reply:
Not exactly, here I want to test case when a source/target table is defined on some node and absent on others.